core/src/main/scala/akka/persistence/r2dbc/internal/sqlserver/SqlServerJournalDao.scala

```diff
@@ -0,0 +1,312 @@
+import akka.persistence.r2dbc.internal.JournalDao
+import akka.persistence.r2dbc.internal.PayloadCodec
+import akka.persistence.r2dbc.internal.R2dbcExecutor
+import akka.persistence.r2dbc.internal.SerializedEventMetadata
+import akka.persistence.r2dbc.internal.Sql.Interpolation
+import akka.persistence.typed.PersistenceId
+import io.r2dbc.spi.Connection
+import io.r2dbc.spi.ConnectionFactory
+import io.r2dbc.spi.Row
+import io.r2dbc.spi.Statement
+import org.slf4j.Logger
+import org.slf4j.LoggerFactory
+
+import java.time.Instant
+import java.time.LocalDateTime
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+/**
+ * INTERNAL API
+ */
+@InternalApi
+private[r2dbc] object SqlServerJournalDao {
+  private val log: Logger = LoggerFactory.getLogger(classOf[SqlServerJournalDao])
+
+  val TRUE = 1
+
+  def readMetadata(row: Row): Option[SerializedEventMetadata] = {
+    row.get("meta_payload", classOf[Array[Byte]]) match {
+      case null => None
+      case metaPayload =>
+        Some(
+          SerializedEventMetadata(
+            serId = row.get[Integer]("meta_ser_id", classOf[Integer]),
+            serManifest = row.get("meta_ser_manifest", classOf[String]),
+            metaPayload))
+    }
+  }
+
+}
+
+/**
+ * INTERNAL API
+ *
+ * Class for doing db interaction outside of an actor to avoid mistakes in future callbacks
+ */
+@InternalApi
+private[r2dbc] class SqlServerJournalDao(settings: R2dbcSettings, connectionFactory: ConnectionFactory)(implicit
+    ec: ExecutionContext,
+    system: ActorSystem[_])
+    extends JournalDao {
+
+  private val helper = SqlServerDialectHelper(settings.connectionFactorySettings.config)
+  import helper._
+
+  require(settings.useAppTimestamp, "SqlServer requires akka.persistence.r2dbc.use-app-timestamp=on")
+  require(settings.useAppTimestamp, "SqlServer requires akka.persistence.r2dbc.db-timestamp-monotonic-increasing = off")
+
+  import JournalDao.SerializedJournalRow
+  protected def log: Logger = SqlServerJournalDao.log
+
+  private val persistenceExt = Persistence(system)
+
+  protected val r2dbcExecutor =
+    new R2dbcExecutor(
+      connectionFactory,
+      log,
+      settings.logDbCallsExceeding,
+      settings.connectionFactorySettings.poolSettings.closeCallsExceeding)(ec, system)
+
+  protected val journalTable = settings.journalTableWithSchema
+  protected implicit val journalPayloadCodec: PayloadCodec = settings.journalPayloadCodec
+
+  private val insertEventWithParameterTimestampSql = sql"""
+    INSERT INTO $journalTable
+    (slice, entity_type, persistence_id, seq_nr, writer, adapter_manifest, event_ser_id, event_ser_manifest, event_payload, tags, meta_ser_id, meta_ser_manifest, meta_payload, db_timestamp)
+    OUTPUT inserted.db_timestamp
+    VALUES (@slice, @entityType, @persistenceId, @seqNr, @writer, @adapterManifest, @eventSerId, @eventSerManifest, @eventPayload, @tags, @metaSerId, @metaSerManifest, @metaSerPayload, @dbTimestamp)"""
```

Comment:
Value binding in this specific driver does not allow positional index binding based on `?`s. The only alternative to what is currently used is positional binding zero-based - so numbers instead of `?`s.

Unless I misunderstand, this means we can not reuse code related to queries definition / binding.

If have no problem with moving to zero-based binding for this PR if you still suggest this is the preferred approach. I  just think named binding is much more readable.

Reply:
For now I kept binding using named param as this seems to be recommended by the plugin: https://github.com/r2dbc/r2dbc-mssql/blob/main/src/main/java/io/r2dbc/mssql/MssqlStatement.java#L29

Reply:
Alright, so you mean that the SqlServer itself uses named parameters in the actual requests sent to the server? That is different to Postgres, which uses positional parameters, and the translation from name to index happens in the client. Maybe the opposite translation from index to name isn't too bad for SqlServer and we could still use the same approach? The advantage of using the same approach would be that much more code would be the same (no need for special bind methods for all sql).

Reply:
Yes, I agree. If the query is similar enough to re-use the binding based on positional index and then using some customization (e.g. for timestamps or tags) that would work out nice. 

I will keep this in mind for later, once we have a direction regarding the other discussion below.